builder/src/upload.tsx

```diff
@@ -0,0 +1,277 @@
+    const [isDragging, setIsDragging] = useState<boolean>(false);
+
+    const resetDragState = () => {
+        setIsDragging(false);
+        setFileDropStyle({
+            height: undefined,
+            border: undefined,
+        });
+    };
+    const windowDragEnter = (e: DragEvent) => {
+        setIsDragging(true);
+        setLastTarget(e.target);
+        if (!props.readonly) {
+            setFileDropStyle({
+                height: "200px",
+                border: "4px solid #fff",
+            });
+        }
+    };
+    const windowDragLeave = (e: DragEvent) => {
+        if (e.target === lastTarget || e.target === document) {
+            resetDragState();
+        }
+    };
+    const windowDrop = () => {
+        resetDragState();
+    };
+    const fileChange = () => {
+        resetDragState();
+    };
+    const onDrop = (e: React.DragEvent) => {
+        if (!props.readonly) {
+            const inp = fileInputRef.current;
+            if (inp) {
+                inp.files = e.dataTransfer.files;
+            }
+            if (props.onChange) {
+                props.onChange(e.dataTransfer.files);
+            }
+        }
+        e.preventDefault();
+        resetDragState();
+    };
+
+    useEffect(() => {
+        window.addEventListener("dragenter", windowDragEnter);
+        window.addEventListener("dragleave", windowDragLeave);
+        window.addEventListener("drop", windowDrop);
+        return () => {
+            window.removeEventListener("dragenter", windowDragEnter);
+            window.removeEventListener("dragleave", windowDragLeave);
+            window.removeEventListener("drop", windowDrop);
+        };
+    });
+
+    const extraClass = !!props.readonly ? "disabled" : "";
+    const finalStyle = {
+        height: fileDropStyle.height,
+        border: fileDropStyle.border,
+        cursor: !!props.readonly ? undefined : "pointer",
+    };
+
+    return (
+        <label
+            className={`btn btn-primary btn-lg btn-block ${extraClass}`}
+            onDragOver={(e) => e.preventDefault()}
+            onDrop={onDrop}
+            style={finalStyle}
+            aria-disabled={props.readonly}
+        >
+            {isDragging && !props.readonly ? "Drag file here" : props.title}
+            <input
+                type="file"
+                name="newfile"
+                style={{ display: "none" }}
+                ref={fileInputRef}
+                onChange={fileChange}
+                disabled={props.readonly}
+            />
+        </label>
+    );
+};
+
+const calculateMD5 = (blob: Blob) => {
+    return new Promise<string>((resolve) => {
+        const reader = new FileReader();
+        reader.readAsBinaryString(blob);
+        reader.onloadend = () => {
+            const md5 = crypto.MD5(
+                crypto.enc.Latin1.parse(reader.result!.toString())
+            );
+            resolve(md5.toString(crypto.enc.Base64));
+        };
+    });
```

Comment:
I don't see how does getting rid of promise chaining on the call site remove the need for this function to return a promise? Returning a promise makes it so we can await an asynchronous task to finish. Reading the binary string happens to be an synchronous task, so we need to wrap it in a promise if we want to wait for it to finish.

Reply:
One thing that we should do is set the callback _before_ loading data to avoid a possible race condition (maybe?), where the loading finishes before the callback is in place. I'm unsure if that's actually a scenario that can happen, but better to prepare for that too if it is.

Reply:
The reason I'm confused about this being async is because it doesn't call any async functions, so what does it actually do? Won't it block regardless? I'm used to the only reason for having an `async` function being if you use `await` in it, otherwise does it actually make any difference compared to it not returning a promise other than added overhead?

Reply:
So, `async` functions always return a promise. You can also have a function return a promise even if the function isn't declared `async`. That just means the promise is built in a blocking fashion, but the _resolution_ of the promise is still asynchronous. You can await any function that returns a promise, as that will just wait for the promise that was returned to resolve. In this case, `readAsBinaryString` is a non-blocking operation, so to wait for it to complete, we return a promise that resolves only after the loading has completed. This is achieved by assigning a function to the `reader.onloadend` variable which resolves the promise, as that's what the reader will call once it has finished loading.

TL:DR
- `async` makes it so a function has to return a promise, and let's you use `await`
- `await` can be used with any function that returns a promise, and will simply wait for the promise to resolve before continuing
- In this case we return a promise because loading the data to the reader is non-blocking. The promise we return is resolved once the loading is complete (thanks to the callback we assign to the reader's onloadend event)

Reply:
I see, when re-reading the code I missed that `FileReader` uses an older callback-based API. I completely understand your point now.

I like the suggestion from [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises#creating_a_promise_around_an_old_callback_api):
> A `Promise` can be created from scratch using its constructor. This should be needed only to wrap old APIs. [...]
>
> Mixing old-style callbacks and promises is problematic. [...]
>
> Best practice is to wrap problematic functions at the lowest possible level, and then never call them directly again

This would result in something like:
```ts
const readAsBinaryString = (blob: Blob) => {
  return new Promise<string | ArrayBuffer>((resolve) => {
    const reader = new FileReader();
    reader.readAsBinaryString(blob);
    reader.onloadend = () => {
      resolve(reader.result);
    };
  });
};

const calculateMD5 = async (blob: Blob) => {
  const content = await readAsBinaryString(blob);
  const md5 = crypto.MD5(crypto.enc.Latin1.parse(content.toString()));
  return md5.toString(crypto.enc.Base64);
};
```

I do prefer this, but after actually writing it, I realise that it's unnecessary considering:
- It won't be called anywhere else
- It's not *that* much nicer than the current implementation as `calculateMD5` is fairly simple excluding `readAsBinaryString` (in the above snippet)